File: src/tools/dump.js
const path = require('path');
const fs = require('fs');
const { PublicKey } = require("@solana/web3.js");

const { createConnection } = require("./tx.js");
const { loadTxFromJson } = require("../tx-format/json-tx.js");

const UPGRADEABLE_LOADER_ID = "BPFLoaderUpgradeab1e11111111111111111111111";
const ELF_MAGIC = Buffer.from([0x7f, 0x45, 0x4c, 0x46]);
const MAX_U64_STR = "18446744073709551615";
const MAX_U64 = BigInt(MAX_U64_STR);

function toBuffer(data) {
    if (!data) return Buffer.alloc(0);
    return Buffer.isBuffer(data) ? data : Buffer.from(data);
}

function extractElfBytes(data) {
    const buf = toBuffer(data);
    if (buf.length === 0) return null;
    const offset = buf.indexOf(ELF_MAGIC);
    if (offset < 0) return null;
    return buf.slice(offset);
}

function tryGetUpgradeableProgramDataAddress(data) {
    const buf = toBuffer(data);
    if (buf.length < 4 + 32) return null;
    const tag = buf.readUInt32LE(0);
    if (tag !== 2) return null;
    try {
        return new PublicKey(buf.slice(4, 36));
    } catch (_) {
        return null;
    }
}

function serializeAccountInfo(pubkey, accountInfo) {
    return {
        pubkey: pubkey.toBase58(),
        account: {
            lamports: accountInfo.lamports,
            data: [toBuffer(accountInfo.data).toString('base64'), 'base64'],
            owner: accountInfo.owner.toBase58(),
            executable: accountInfo.executable,
            rentEpoch: normalizeU64String(accountInfo.rentEpoch),
            space: toBuffer(accountInfo.data).length,
        },
    };
}

function normalizeU64String(value) {
    if (typeof value === 'bigint') {
        if (value < 0n) return "0";
        return value > MAX_U64 ? MAX_U64_STR : value.toString();
    }
    if (typeof value === 'number') {
        if (!Number.isFinite(value) || value < 0) return "0";
        if (value > Number.MAX_SAFE_INTEGER) {
            return MAX_U64_STR;
        }
        return Math.floor(value).toString();
    }
    if (typeof value === 'string') {
        if (!/^\d+$/.test(value)) return "0";
        try {
            const big = BigInt(value);
            if (big < 0n) return "0";
            return big > MAX_U64 ? MAX_U64_STR : value;
        } catch (_) {
            return "0";
        }
    }
    return "0";
}

function stringifyAccountPayload(payload) {
    const json = JSON.stringify(payload, null, '\t');
    return json.replace(/"rentEpoch":\s*"(\d+)"/g, '"rentEpoch": $1');
}

function normalizeAccountKey(account) {
    if (!account) return null;
    if (typeof account === 'string') return account;
    if (account.pubkey) {
        if (typeof account.pubkey === 'string') return account.pubkey;
        if (typeof account.pubkey.toBase58 === 'function') return account.pubkey.toBase58();
        if (typeof account.pubkey.toString === 'function') return account.pubkey.toString();
    }
    if (typeof account.toBase58 === 'function') return account.toBase58();
    if (typeof account.toString === 'function') return account.toString();
    return null;
}

function addAccountsFromList(targetSet, accounts) {
    if (!Array.isArray(accounts)) return;
    accounts.forEach((account) => {
        const key = normalizeAccountKey(account);
        if (key) {
            targetSet.add(key);
        }
    });
}

async function dumpAccount(address, toPath) {
    if (!fs.existsSync(toPath)) {
        fs.mkdirSync(toPath, { recursive: true });
    }

    const connection = createConnection('http://api.mainnet-beta.solana.com');
    const pubkey = new PublicKey(address);
    const accountInfo = await connection.getAccountInfo(pubkey);
    if (!accountInfo) {
        throw new Error(`Account not found: ${address}`);
    }
    if (accountInfo.executable) {
        console.log(`Dumping program ${address}...`);
        let programData = accountInfo.data;
        if (accountInfo.owner?.toBase58() === UPGRADEABLE_LOADER_ID) {
            const programDataAddress = tryGetUpgradeableProgramDataAddress(accountInfo.data);
            if (programDataAddress) {
                const programDataInfo = await connection.getAccountInfo(programDataAddress);
                if (programDataInfo?.data) {
                    programData = programDataInfo.data;
                }
            }
        }

        const elfBytes = extractElfBytes(programData);
        if (!elfBytes || elfBytes.length === 0) {
            throw new Error(`Program data not found or not ELF for: ${address}`);
        }

        const outPath = path.join(toPath, `${address}.so`);
        fs.writeFileSync(outPath, elfBytes);
        console.log(`Program dumped to ${outPath}`);
    } else {
        console.log(`Dumping account ${address}...`);
        const outPath = path.join(toPath, `${address}.json`);
        const payload = serializeAccountInfo(pubkey, accountInfo);
        fs.writeFileSync(outPath, stringifyAccountPayload(payload));
        console.log(`Account dumped to ${outPath}`);
    }
}

async function dumpAccountsFromTx(signature, toPath) {
    const connection = createConnection('http://api.mainnet-beta.solana.com');
    const tx = await connection.getTransaction(signature, { commitment: 'confirmed', maxSupportedTransactionVersion: 0 });
    if (!tx) {
        throw new Error(`Transaction not found: ${signature}`);
    }

    const accounts = new Set();
    const message = tx.transaction?.message;
    addAccountsFromList(accounts, message?.accountKeys);
    addAccountsFromList(accounts, message?.staticAccountKeys);
    addAccountsFromList(accounts, tx.meta?.loadedAddresses?.writable);
    addAccountsFromList(accounts, tx.meta?.loadedAddresses?.readonly);

    const tokenBalances = [
        ...(tx.meta?.preTokenBalances ?? []),
        ...(tx.meta?.postTokenBalances ?? []),
    ];
    const tokenMints = tokenBalances.map((balance) => balance?.mint).filter(Boolean);
    const tokenOwners = tokenBalances.map((balance) => balance?.owner).filter(Boolean);
    addAccountsFromList(accounts, tokenMints);
    addAccountsFromList(accounts, tokenOwners);

    for (const account of accounts) {
        try {
            await dumpAccount(account, toPath);
        } catch (error) {
            console.error(`Failed to dump account ${account}:`, error.message);
        }
    }
}

async function dumpAccountsForTx(path, toPath, params = []) {
    const tx = loadTxFromJson(path, params);

    const accounts = new Set();
    tx.forEach(instruction => {
        instruction.accounts.forEach(account => {
            accounts.add(account.pubkey.toBase58());
        });
    });

    for (const account of accounts) {
        try {
            await dumpAccount(account, toPath);
        } catch (error) {
            console.error(`Failed to dump account ${account}:`, error.message);
        }
    }
}

async function dumpRawTransaction(signature, toPath = '.') {
    const connection = createConnection('http://api.mainnet-beta.solana.com');
    const tx = await connection.getParsedTransaction(signature, { commitment: 'confirmed', maxSupportedTransactionVersion: 0 });
    if (!tx) {
        throw new Error(`Transaction not found: ${signature}`);
    }

    if (!fs.existsSync(toPath)) {
        fs.mkdirSync(toPath, { recursive: true });
    }

    const filePath = path.join(toPath, `${signature}.json`);
    fs.writeFileSync(filePath, JSON.stringify(tx, null, '\t'));
    console.log(`Raw transaction dumped to ${filePath}`);
}

async function dumpRawBlock(slot, toPath = '.') {
    const connection = createConnection('http://api.mainnet-beta.solana.com');
    const blockNumber = Number(slot);
    if (Number.isNaN(blockNumber)) {
        throw new Error(`Invalid slot: ${slot}`);
    }

    const block = await connection.getParsedBlock(blockNumber, { commitment: 'confirmed', maxSupportedTransactionVersion: 0 });
    if (!block) {
        throw new Error(`Block not found: ${slot}`);
    }

    if (!fs.existsSync(toPath)) {
        fs.mkdirSync(toPath, { recursive: true });
    }

    const filePath = path.join(toPath, `${blockNumber}.json`);
    fs.writeFileSync(filePath, JSON.stringify(block, null, '\t'));
    console.log(`Raw block dumped to ${filePath}`);
}

module.exports = {
    dumpAccount,
    dumpAccountsFromTx,
    dumpAccountsForTx,
    dumpRawTransaction,
    dumpRawBlock,
};

File: src/tools/dump.rs
use std::{collections::HashSet, fs, path::Path, str::FromStr};

use anyhow::{Context, Result, anyhow};
use base64::{Engine as _, engine::general_purpose::STANDARD};
use solana_commitment_config::CommitmentConfig;
use solana_rpc_client::api::config::RpcTransactionConfig;
use solana_sdk::pubkey::Pubkey;
use solana_transaction_status::{EncodedTransaction, UiMessage, UiTransactionEncoding};

use crate::tools::tx::{MAINNET_RPC_URL, create_connection};
use crate::tx_format::json_tx::load_parsed_tx_from_json;

const UPGRADEABLE_LOADER_ID: Pubkey =
    Pubkey::from_str_const("BPFLoaderUpgradeab1e11111111111111111111111");
const ELF_MAGIC: [u8; 4] = [0x7f, 0x45, 0x4c, 0x46];

fn extract_elf_bytes(data: &[u8]) -> Option<Vec<u8>> {
    data.windows(ELF_MAGIC.len())
        .position(|window| window == ELF_MAGIC)
        .map(|idx| data[idx..].to_vec())
}

fn try_get_upgradeable_program_data_address(data: &[u8]) -> Option<Pubkey> {
    if data.len() < 4 + 32 {
        return None;
    }
    let tag = u32::from_le_bytes(data[0..4].try_into().ok()?);
    if tag != 2 {
        return None;
    }
    Some(Pubkey::new_from_array(data[4..36].try_into().ok()?))
}

fn serialize_account_info(
    pubkey: &Pubkey,
    account: &solana_sdk::account::Account,
) -> serde_json::Value {
    serde_json::json!({
        "pubkey": pubkey.to_string(),
        "account": {
            "lamports": account.lamports,
            "data": [STANDARD.encode(&account.data), "base64"],
            "owner": account.owner.to_string(),
            "executable": account.executable,
            "rentEpoch": account.rent_epoch,
            "space": account.data.len(),
        }
    })
}

pub fn dump_account(address: &str, to_path: impl AsRef<Path>) -> Result<()> {
    fs::create_dir_all(&to_path)?;

    let connection = create_connection(MAINNET_RPC_URL);
    let pubkey = Pubkey::from_str(address).map_err(|_| anyhow!("Invalid pubkey: {address}"))?;
    let account = connection
        .get_account(&pubkey)
        .with_context(|| format!("Account not found: {address}"))?;

    if account.executable {
        println!("Dumping program {address}...");
        let mut program_data = account.data.clone();
        if account.owner == UPGRADEABLE_LOADER_ID {
            if let Some(program_data_address) =
                try_get_upgradeable_program_data_address(&account.data)
            {
                if let Ok(program_data_info) = connection.get_account(&program_data_address) {
                    program_data = program_data_info.data;
                }
            }
        }

        let elf_bytes = extract_elf_bytes(&program_data)
            .ok_or_else(|| anyhow!("Program data not found or not ELF for: {address}"))?;
        let out_path = to_path.as_ref().join(format!("{address}.so"));
        fs::write(&out_path, elf_bytes)?;
        println!("Program dumped to {}", out_path.display());
    } else {
        println!("Dumping account {address}...");
        let payload = serialize_account_info(&pubkey, &account);
        let out_path = to_path.as_ref().join(format!("{address}.json"));
        fs::write(&out_path, serde_json::to_string_pretty(&payload)?)?;
        println!("Account dumped to {}", out_path.display());
    }

    Ok(())
}

fn add_account(set: &mut HashSet<String>, account: &str) {
    if !account.is_empty() {
        set.insert(account.to_string());
    }
}

pub fn dump_accounts_from_tx(signature: &str, to_path: impl AsRef<Path>) -> Result<()> {
    let connection = create_connection(MAINNET_RPC_URL);
    let config = RpcTransactionConfig {
        encoding: Some(UiTransactionEncoding::JsonParsed),
        commitment: Some(CommitmentConfig::confirmed()),
        max_supported_transaction_version: Some(0),
    };
    let tx = connection
        .get_transaction_with_config(&signature.parse()?, config)
        .with_context(|| format!("Transaction not found: {signature}"))?;

    let mut accounts = HashSet::new();
    let message = match &tx.transaction.transaction {
        EncodedTransaction::Json(tx) => &tx.message,
        _ => return Err(anyhow!("Transaction encoding is not JSON")),
    };

    match message {
        UiMessage::Parsed(msg) => {
            for key in &msg.account_keys {
                add_account(&mut accounts, &key.pubkey);
            }
        }
        UiMessage::Raw(msg) => {
            for key in &msg.account_keys {
                add_account(&mut accounts, key);
            }
        }
    }

    if let Some(meta) = tx.transaction.meta {
        let loaded_addresses: Option<solana_transaction_status::UiLoadedAddresses> =
            meta.loaded_addresses.into();
        if let Some(loaded) = loaded_addresses {
            for key in loaded.writable.iter().chain(loaded.readonly.iter()) {
                add_account(&mut accounts, key);
            }
        }

        let token_balances: Vec<solana_transaction_status::UiTransactionTokenBalance> =
            Option::<Vec<_>>::from(meta.pre_token_balances)
                .unwrap_or_default()
                .into_iter()
                .chain(
                    Option::<Vec<_>>::from(meta.post_token_balances)
                        .unwrap_or_default()
                        .into_iter(),
                )
                .collect();

        for balance in token_balances {
            add_account(&mut accounts, &balance.mint);
            if let Some(owner) = Option::<String>::from(balance.owner) {
                add_account(&mut accounts, &owner);
            }
        }
    }

    for account in accounts {
        if let Err(error) = dump_account(&account, &to_path) {
            eprintln!("Failed to dump account {account}: {error}");
        }
    }

    Ok(())
}

pub fn dump_accounts_for_tx(
    path: impl AsRef<Path>,
    to_path: impl AsRef<Path>,
    params: &[String],
) -> Result<()> {
    let tx = load_parsed_tx_from_json(&path, params)?;

    let mut accounts = HashSet::new();
    for instruction in tx.instructions {
        for account in instruction.accounts {
            add_account(&mut accounts, &account.pubkey.to_string());
        }
    }

    for account in accounts {
        if let Err(error) = dump_account(&account, &to_path) {
            eprintln!("Failed to dump account {account}: {error}");
        }
    }

    Ok(())
}

pub fn dump_raw_transaction(signature: &str, to_path: impl AsRef<Path>) -> Result<()> {
    let connection = create_connection(MAINNET_RPC_URL);
    let config = RpcTransactionConfig {
        encoding: Some(UiTransactionEncoding::JsonParsed),
        commitment: Some(CommitmentConfig::confirmed()),
        max_supported_transaction_version: Some(0),
    };
    let tx = connection
        .get_transaction_with_config(&signature.parse()?, config)
        .with_context(|| format!("Transaction not found: {signature}"))?;

    fs::create_dir_all(&to_path)?;
    let file_path = to_path.as_ref().join(format!("{signature}.json"));
    fs::write(&file_path, serde_json::to_string_pretty(&tx)?)?;
    println!("Raw transaction dumped to {}", file_path.display());
    Ok(())
}

pub fn dump_raw_block(slot: &str, to_path: impl AsRef<Path>) -> Result<()> {
    let connection = create_connection(MAINNET_RPC_URL);
    let slot_num: u64 = slot.parse().map_err(|_| anyhow!("Invalid slot: {slot}"))?;

    let config = solana_rpc_client::api::config::RpcBlockConfig {
        encoding: Some(UiTransactionEncoding::JsonParsed),
        transaction_details: Some(solana_transaction_status::TransactionDetails::Full),
        rewards: Some(true),
        commitment: Some(CommitmentConfig::confirmed()),
        max_supported_transaction_version: Some(0),
    };

    let block = connection
        .get_block_with_config(slot_num, config)
        .with_context(|| format!("Block not found: {slot}"))?;

    fs::create_dir_all(&to_path)?;
    let file_path = to_path.as_ref().join(format!("{slot_num}.json"));
    fs::write(&file_path, serde_json::to_string_pretty(&block)?)?;
    println!("Raw block dumped to {}", file_path.display());
    Ok(())
}
